### PR TITLE
[WFLY-16527] Faces 4 fails if CDI isn't enabled for the deployment

### PR DESCRIPTION
**Pull request by jasondlee**
https://issues.redhat.com/browse/WFLY-16527

Add check to require CDI for Faces deployments
Add method to WeldCapability to force Weld initialization
Call new method to force CDI init for Faces deployments

### CODE REVIEW

File: weld/subsystem/src/main/java/org/jboss/as/weld/WeldCapabilityImpl.java
Comment (bstansberry) [inline weld/subsystem/src/main/java/org/jboss/as/weld/WeldCapabilityImpl.java:78]: This should call WeldDeploymentMarker.mark

Comment (bstansberry) [reply weld/subsystem/src/main/java/org/jboss/as/weld/WeldCapabilityImpl.java:78]: I guess this way can work too, but it seems cleaner to just use the existing WeldDeploymentMarker whose purpose is to track this state.

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFComponentProcessor.java
Comment (bstansberry) [inline jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFComponentProcessor.java:260]: The JSFResourceDefinition should declare an 'org.wildfly.faces' capability that requires WELD_CAPABILITY_NAME. Convert the JSFResourceDefinition c'tor to pass a SimpleResourceDefinition.Parameters object to super(...) and pass that new capability into that object.

Once you have done that there is no longer any possibility that WeldCapability will not be present when a jsf subsystem DUP runs.

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFComponentProcessor.java
Comment (bstansberry) [inline jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFComponentProcessor.java:303]: Yep; isCDI isn't needed any longer. Any other jsf subsystem code that is conditionally checking for CDI can stop too once JSFResourceDefinition declare an 'org.wildfly.faces' capability that requires WELD_CAPABILITY_NAME.

**Review by bstansberry — CHANGES_REQUESTED** [review]
See my previous comments for requested changes.

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFResourceDefinition.java
Comment (jamezp) [inline jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFResourceDefinition.java:78]: I assume this was the example and can be removed :)

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFResourceDefinition.java
Comment (jamezp) [inline jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFResourceDefinition.java:88]: This can likely be removed too.

**Review by jamezp — COMMENTED** [review]
Just two minor comments, but otherwise looks good to me.

**Comment by jamezp** [comment]
Actually, the test failures seem to be related.

File: jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFResourceDefinition.java
Comment (jasondlee) [reply jsf/subsystem/src/main/java/org/jboss/as/jsf/subsystem/JSFResourceDefinition.java:78]: Good grief. Yes, it can. :P

**Review by bstansberry — CHANGES_REQUESTED** [review]
@jasondlee The functional change looks good.

Can you remove this line please? 

https://github.com/wildfly/wildfly/blob/main/testsuite/integration/basic/pom.xml#L3538

I believe with -Dts.ee9 that test should now pass, and if it doesn't that may mean something more is needed.

My guess is passing AdditionalInitialization.MANAGEMENT instead of null in this line will fix the subsystem test:

https://github.com/wildfly/wildfly/blob/main/jsf/subsystem/src/test/java/org/jboss/as/jsf/subsystem/JSFSubsystemTestCase.java#L58

You've added a subsystem requirement for WeldCapability. For that to work you need to provide the capability, unless the test controller starts in admin-only mode. The subsystem test framework has hooks for providing capabilities, but in this case starting in admin-only is sufficient to check what the test checks, so doing that is the simplest solution. Passing AdditionalInitialization.MANAGEMENT does that.

If you make those changes and it passes CI, please don't wait for my approval to get this merged, as I may not check back for some days.

**Comment by jasondlee** [comment]
I've updated the POM and the test change suggestion worked. I'll push the changes, then find someone to merge when it's green. Thanks!

**Comment by bstansberry** [comment]
@jasondlee Once this is merged and a 'main' with this in it is merged into the big-bang branch, can you please send up a PR to big-bang removing this line?

https://github.com/wildfly/wildfly/blob/big-bang/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java#L55

Thanks!

**Comment by jasondlee** [comment]
> @jasondlee Once this is merged and a 'main' with this in it is merged into the big-bang branch, can you please send up a PR to big-bang removing this line?

Prepped and ready to go: https://github.com/wildfly/wildfly/pull/15725

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java
Comment (jasondlee) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java:113]: @jamezp @bstansberry, This hack is a bit ugly, but, other than continuing to ignore the test, I'm not sure how best to go about it. That said, the test does now pass in main and WFP as expected.  Also, IDEA cleaned up some formatting and import order that I didn't notice before pushing. I don't mind the clean up, but if you'd like less noise, I can reset and make a quieter change for this. :)

Comment (jamezp) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java:113]: I assume `javax.faces.resource` has special meeting in which case I can't think of a better way to handle it.

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java
Comment (bstansberry) [inline testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java:125]: Instead of checking the active profile, which will break for other profiles or in big-bang, please inspect the FQCN of a Faces API class. If it starts with 'jakarta' then the test is in a 'jakarta' environment.

Comment (jasondlee) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java:125]: This is a client test (@RunAsClient), so the namespace is not being transformed, giving me a javax.faces.context.FacesContext in the test method. Class.forName("jakarta.faces.context.FacesContext") throws a CNFE. Am I missing a trick?

Comment (bstansberry) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java:125]: Ah, ok, sorry; then testing for the profile is ok. Please use AssumeTestGroupUtil.isWildFlyPreview() though.

#15725 can back this out, as the class already uses the jakarta.* URL in the big-bang branch. That's just minor housekeeping though as the method will still work, it just won't do anything.

**Comment by bstansberry** [comment]
@darranl If you want to merge this as-is on Monday, go ahead. My last comment above is a minor thing.

**Review by fjuma — APPROVED** [review]
@darranl Sorry for the delayed response, it was a holiday here on Friday. This looks good to me.

**Comment by darranl** [comment]
No problem, thank you @fjuma I remembered just as I pinged you

File: testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java
Comment (jasondlee) [reply testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jsf/testurl/ForbiddenUrlTestCase.java:116]: It is. It's a mean to access bundled resources in META-INF, iirc.

**Comment by jasondlee** [comment]
> > @jasondlee Once this is merged and a 'main' with this in it is merged into the big-bang branch, can you please send up a PR to big-bang removing this line?
> 
> Prepped and ready to go: #15725

I'll update this pending PR to remove the WFP check once the change here is merged.

**Pull request merged**
